libs/langgraph/tests/test_pregel.py

```diff
@@ -5491,3 +5491,241 @@ def invoke_sub_agent(state: AgentState):
             "invoke_sub_agent": {"input": True},
         },
     ]
+
+
+def test_map_reduce() -> None:
+    """Test map reduce with Graph API."""
+    import operator
+    from typing import Annotated, NotRequired, TypedDict
+
+    class OverallState(TypedDict):
+        subjects: list
+        # Notice here we use the operator.add
+        # This is because we want combine all the jokes we generate
+        # from individual nodes back into one list - this is essentially
+        # the "reduce" part
+        jokes_with_grades: Annotated[list, operator.add]
+        best_selected_joke: str
+
+    # This will be the state of the node that we will "map" all
+    # subjects to in order to generate a joke
+    class JokeState(TypedDict):
+        subject: str
+        joke: NotRequired[str]
+        joke_grade: NotRequired[int]
+
+    def generate_jokes(state: OverallState):
+        # Distribute the subjects to the generate_joke node
+        return [Send("generate_joke", {"subject": s}) for s in state["subjects"]]
+
+    def generate_joke(state: JokeState):
+        # Come up with a joke and continue to grade the joke
+        funny_joke = f"Funny joke about {state['subject']}"
+        return [Send("grade_joke", {"joke": funny_joke, "subject": state["subject"]})]
+
+    def grade_joke(state: JokeState):
+        # Grade the joke and then send it to the count_grade node which will reduce
+        if state["subject"].startswith("a"):
+            grade = 10
+        elif state["subject"].startswith("b"):
+            grade = 9
+        else:
+            grade = 1
+        return {
+            "jokes_with_grades": [
+                {
+                    "joke": state["joke"],
+                    "grade": grade,
+                }
+            ]
+        }
+
+    def count_grade(state: OverallState):
+        max_score = 0
+        best_joke = ""
+        for joke in state["jokes_with_grades"]:
+            if joke["grade"] > max_score:
+                max_score = joke["grade"]
+                best_joke = joke["joke"]
+        return {"best_selected_joke": best_joke}
+
+    builder = StateGraph(OverallState)
+    builder.add_conditional_edges("__start__", generate_jokes, ["generate_joke"])
```

Comment:
fails compilation:

>               raise ValueError(f"Found edge starting at unknown node '{source}'")
E               ValueError: Found edge starting at unknown node 'generate_joke'


1) This should probably work
2) We should've called these "add_routing_node" in retrospect

Reply:
But your graph doesn't have a "generate_joke" node? What am I missing?

Reply:
If i add a node it requires the state to be overall state (and I think forces the reduce step)